POEM_003.md

```diff
@@ -0,0 +1,111 @@
+Furthermore, the communicators are not created until `setup` is called, therefore users cannot initialize supplementary Python objects that require access to the communicator.
+
+Frequently, these objects that require the communicator needs to be initialized to determine the input and output sizes for their components, effectively creating a circular dependency.
+This factor makes it difficult to integrate complied analysis codes in OpenMDAO components in a modular way.
+This problem is currently avoided by ad hoc implementations, where the Python objects pass information in a layer hidden from OpenMDAO, which is not ideal and limits the flexibility of the framework.
+
+
+
+
+Suggested solution
+==================
+
+We suggest adding a `pre_setup` method to the `Problem` class in OpenMDAO.
+This method will:
+
+1. Create the communicators,
+2. traverse down the model hierarchy and call the `pre_setup` method on all components, and
+3. traverse up the model hierarchy and call the `pre_configure` method on all components.
+
+After `pre_setup` is called, the user will still be able to make new connections in the model, and define new input and outputs for any component.
+Furthermore, the user will be able to access the communicators, and the objects created for each component as attributes of the OpenMDAO `problem` instance.
+The model hierarchy can be fixed after this step.
+We do not have any strict requirements on any of the method names. The 3rd step is open to discussion, however, our main suggestion is adding step 2.
+
+Our goal with this approach is to define a way to initialize the communicators and Python objects, before any of the input and output vector sizes are frozen.
+This enables compiled analysis codes that require access to the communicators to be initialized and saved as attributes of their OpenMDAO components that contain them.
+
+Because the user can still add new inputs and outputs, and make new connections after this stage, the user can access the created Python objects and communicate across components at the top level script.
+This enables mimicking the Python-based optimization frameworks such as the MACH framework developed by the MDO Lab.
+
+Example script
+==============
+
+The pseudo-code for an example optimization script is listed below:
+
+
+```python
+from openmdao.api import Problem
+from MACH-Aero import DVGeometryComp, ADflowComp
+
+# initialize the OpenMDAO problem
+p = Problem()
+
+# add geometry component
+p.model.add_subsystem('geometry', DVGeometryComp())
+
+# add solver component
+p.model.add_subsystem('solver', ADflowComp())
+
+# call pre_setup
+# all communicators and components are initialized with this method
+p.pre_setup()
+
+# add geometry design variables
+# this method also adds 'twist' as an output of the comp
+p.model.geometry.addDVGeoGlobal('twist')
+
+# get surface coordinates from solver
+surfaceCoords = p.model.solver.getSurfaceCoordinates()
```

Comment:
This is a core question that I want to understand better, before I can comment more fully on this POEM: 

In lines 76 and 79, you're calling methods on the component or group instances that you created above. How critical do you feel that it is that these methods be called on directly on the OpenMDAO derived sub-classes themselves, instead of on delegate objects that reside in those components. In other words, could you call those methods on generic python classes that get passed into the OpenMDAO component/groups? 

At my first reading these two calls seem like solver/geometry tool specific calls that are not necessarily general to all solvers or geometry tools.

Reply:
These calls are not solver specific and they define an API that enables the use of a geometry tool with a solver. We can define other interfaces, but the root problem persists; we need to be able to take the initial aerodynamic problem surface definition and add it as an output to the geometry tool, which will be upstream of the aerodynamic solver.